### PR TITLE
Add copy deck route

### LINKED ISSUE #117: Add copy deck logic
**User story:**
As a user, I want to start with an existing deck and customize it to my needs.

**Description:**
I would suggest sending the target groupID the following endpoint:  `/api/decks/:deckID/copy`

### PR DESCRIPTION
**Pull request by TomRomeo**
## Description
This PR implements the backend logic to copy decks

## Resolved Issues
resolves #117

### CODE REVIEW

File: backend/services/carddeck/handler/carddeck.go [hunk 1]
```diff
@@ -217,13 +217,68 @@ func (e *CardDeck) CreateDeck(ctx context.Context, req *pb.CreateDeckRequest, rs
...
+		return err
+	}
+	var (
+		deckType model.DeckType
+		err      error
+	)
+	if req.DeckType == nil {
+		deckType = model.PrivateDeckType
+	} else {
+		err, deckType = converter.MigrateProtoDeckTypeToModelDeckType(*req.DeckType)
```
Comment (4KevR) [inline backend/services/carddeck/handler/carddeck.go:235]: Please refactor method to first return the `deckType` and then the `err` to make it consistent with the existing codebase.

File: backend/services/carddeck/handler/carddeck.go [hunk 2]
```diff
@@ -217,13 +217,68 @@ func (e *CardDeck) CreateDeck(ctx context.Context, req *pb.CreateDeckRequest, rs
...
+			pbCardSides = append(pbCardSides, &pb.CardSideContent{
+				Header:      cardSide.Header,
+				Description: cardSide.Description,
+			})
+		}
+		if err := e.generateCardSidesForCard(*newCard, pbCardSides); err != nil {
+			return err
+		}
+	}
+	return nil
```
Comment (4KevR) [inline backend/services/carddeck/handler/carddeck.go:272]: `rsp` of type `DeckResponse` is not used. Suggestion: Use the `IDResponse` or `SuccessResponse`, depending on the needs of the frontend.

**Review by 4KevR — CHANGES_REQUESTED** [review]
Looks great, just some small nits.

File: backend/services/carddeck/handler/carddeck.go (hunk 1 above)
Comment (darmiel) [inline backend/services/carddeck/handler/carddeck.go:235]: swap this if-statement

```suggestion
	if req.DeckType != nil {
		deckType, err = converter.MigrateProtoDeckTypeToModelDeckType(*req.DeckType)
	} else {
		deckType = model.PrivateDeckType
```

File: backend/services/frontend/handler/frontend.go [hunk 3]
```diff
@@ -576,6 +576,30 @@ func (e *Frontend) ModifyDeckHandler(c *fiber.Ctx) error {
 	return c.SendStatus(200)
 }
 
+func (e *Frontend) CopyDeckHandler(c *fiber.Ctx) error {
+	var data map[string]string
```
Comment (darmiel) [inline backend/services/frontend/handler/frontend.go:580]: create a struct for this

File: backend/services/frontend/handler/frontend.go [hunk 4]
```diff
@@ -576,6 +576,30 @@ func (e *Frontend) ModifyDeckHandler(c *fiber.Ctx) error {
...
+	if err := c.BodyParser(&data); err != nil {
+		return err
+	}
+	if deckName := data["deckName"]; deckName == "" {
```
Comment (darmiel) [inline backend/services/frontend/handler/frontend.go:584]: there is no need to create this scoped `deckName` variable. just do

```go
if data["deckName"] == "" {
```

(after the struct change):

```go
if payload.DeckName == "" {
```

**Review by darmiel — COMMENTED** [review]
awsom

File: backend/services/carddeck/handler/carddeck.go (hunk 1 above)
Comment (TomRomeo) [reply backend/services/carddeck/handler/carddeck.go]: done in 2b00ace

File: backend/services/carddeck/handler/carddeck.go [hunk 5]
```diff
@@ -217,13 +217,68 @@ func (e *CardDeck) CreateDeck(ctx context.Context, req *pb.CreateDeckRequest, rs
...
+		if err != nil {
+			return err
+		}
+	}
+	newDeck := &model.Deck{
+		GroupID:  req.TargetGroupID,
+		Name:     req.DeckName,
+		DeckType: deckType,
+	}
+	e.store.CreateDeck(newDeck)
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/services/carddeck/handler/carddeck.go]: handled in 2b00ace

File: backend/services/carddeck/handler/carddeck.go (hunk 2 above)
Comment (TomRomeo) [reply backend/services/carddeck/handler/carddeck.go:272]: changed to `IDResponse` in 2b00ace

File: backend/services/frontend/handler/frontend.go [hunk 6]
```diff
@@ -576,6 +576,30 @@ func (e *Frontend) ModifyDeckHandler(c *fiber.Ctx) error {
...
+	_, err := e.cardDeckService.CopyDeck(c.Context(), &pbCardDeck.CopyDeckRequest{
+		UserID:        userID,
+		DeckID:        c.Params("deckID"),
+		TargetGroupID: data["targetGroupID"],
+		DeckName:      data["deckName"],
+	})
+	if err != nil {
+		return err
+	}
+	return c.SendStatus(200)
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/services/frontend/handler/frontend.go]: done in 2b00ace

File: backend/services/carddeck/handler/carddeck.go (hunk 1 above)
Comment (TomRomeo) [reply backend/services/carddeck/handler/carddeck.go]: implemented in 2b00ace

File: backend/services/frontend/handler/frontend.go (hunk 4 above)
Comment (TomRomeo) [reply backend/services/frontend/handler/frontend.go]: done in 2b00ace

**Review by 4KevR — APPROVED** [review]

**Review by darmiel — APPROVED** [review]

File: backend/services/carddeck/handler/carddeck.go [hunk 7]
```diff
@@ -217,13 +242,53 @@ func (e *CardDeck) CreateDeck(ctx context.Context, req *pb.CreateDeckRequest, rs
 	return nil
 }
 
+func (e *CardDeck) CopyDeck(ctx context.Context, req *pb.CopyDeckRequest, rsp *pb.IDResponse) error {
+	logger.Infof("Received CardDeck.CopyDeck request: %v", req)
+	if err := e.checkUserDeckAccess(ctx, req.UserID, req.DeckID); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/services/carddeck/handler/carddeck.go:249]: ✅

**Review by memeToasty — APPROVED** [review]

**Review by Julian702 — CHANGES_REQUESTED** [review]
Seems to work as far as I tested it, just found one problem:

- endpoint `/api/groups/:groupID/decks` returns an empty string as `groupID`for every deck

File: backend/api-documentation/decks/copy deck.bru [hunk 8]
```diff
@@ -0,0 +1,22 @@
+meta {
+  name: copy deck
+  type: http
+  seq: 9
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/api-documentation/decks/copy deck.bru]: changed in 22ef430

File: backend/api-documentation/decks/copy deck.bru [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+auth:bearer {
+  token: {{accessToken}}
+}
+
+body:json {
+  {
+      "targetGroupID": "G-12345678",
+      "deckName": "test group"
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/api-documentation/decks/copy deck.bru]: changed in 22ef430

File: backend/api-documentation/decks/copy deck.bru [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  }
+}
```
[the remark this answers is not in the document]
Comment (TomRomeo) [reply backend/api-documentation/decks/copy deck.bru:22]: changed in 22ef430

File: backend/api-documentation/decks/copy deck.bru (hunk 8 above)
[the remark this answers is not in the document]
Comment (Julian702) [reply backend/api-documentation/decks/copy deck.bru]: its not changed

**Review by Julian702 — APPROVED** [review]

**Pull request merged**